crates/iota-keys/src/keystore.rs

```diff
@@ -175,9 +197,101 @@ pub struct Alias {
     pub public_key_base64: String,
 }
 
+#[expect(clippy::large_enum_variant)]
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(
+    tag = "type",            // this makes {"type": "...", "value": …}
+    content = "value",       // name the payload field "value"
+    rename_all = "snake_case"
+)]
+pub enum StoredKey {
+    #[serde(with = "serde_iota_keypair")]
+    KeyPair(IotaKeyPair),
+    External {
+        source: String,
+        #[serde(
+            with = "serde_derivation_path::option",
+            default,
+            skip_serializing_if = "Option::is_none"
+        )]
+        derivation_path: Option<DerivationPath>,
+        public_key_base64: String,
+    },
+}
+
+impl Clone for StoredKey {
+    fn clone(&self) -> Self {
+        match self {
+            StoredKey::KeyPair(keypair) => StoredKey::KeyPair(keypair.copy()),
+            StoredKey::External {
+                source,
+                derivation_path,
+                public_key_base64,
+            } => StoredKey::External {
+                source: source.clone(),
+                derivation_path: derivation_path.clone(),
+                public_key_base64: public_key_base64.clone(),
+            },
+        }
+    }
+}
+
+impl From<IotaKeyPair> for StoredKey {
+    fn from(keypair: IotaKeyPair) -> Self {
+        StoredKey::KeyPair(keypair)
+    }
+}
+
+impl StoredKey {
+    pub fn address(&self) -> IotaAddress {
+        IotaAddress::from(&self.public())
+    }
+
+    pub fn public(&self) -> PublicKey {
+        match self {
+            StoredKey::KeyPair(keypair) => keypair.public(),
+            StoredKey::External {
+                public_key_base64, ..
+            } => PublicKey::decode_base64(public_key_base64).unwrap(),
+        }
+    }
+
+    pub fn public_base64(&self) -> String {
+        match self {
+            StoredKey::KeyPair(keypair) => keypair.public().encode_base64(),
+            StoredKey::External {
+                public_key_base64, ..
+            } => public_key_base64.clone(),
+        }
+    }
+
+    pub fn derivation_path(&self) -> Option<DerivationPath> {
+        match self {
+            StoredKey::KeyPair(_) => None,
+            StoredKey::External {
+                derivation_path, ..
+            } => derivation_path.clone(),
+        }
+    }
+
+    pub fn external_source(&self) -> Option<String> {
+        match self {
+            StoredKey::KeyPair(_) => None,
+            StoredKey::External { source, .. } => Some(source.clone()),
+        }
+    }
+
+    pub fn try_keypair(&self) -> Result<&IotaKeyPair, anyhow::Error> {
```

Comment:
Can we rename this to `try_as_keypair` or just `as_keypair`?

Reply:
Renamed to `as_keypair` in https://github.com/iotaledger/iota/pull/7704/commits/62feef7bd7be78d9c04c241d31d1bc3ce062d8d9